### DISCUSSION TITLE
Make PyObject subclass Generic & expose it to users

### DISCUSSION THREAD

**Discussion by rsayn** [pydantic/pydantic#3833] [Feature Request]
Hello,
I was thinking that we could add type hinting to `PyObject` in order to enable editor suggestions based on the actual runtime object that is passed or imported.

I experimented a bit locally and I came up with the following code:

```python
# inside types.py
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ PYOBJECT TYPE ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


class _PyObject(Generic[T]):
    validate_always = True

    @classmethod
    def __get_validators__(cls) -> 'CallableGenerator':
        yield cls.validate

    @classmethod
    def validate(cls, value: Any) -> T:
        if callable(value):
            return cast(T, value)

        try:
            value = str_validator(value)
        except errors.StrError:
            raise errors.PyObjectError(error_message='value is neither a valid import path not a valid callable')

        try:
            return cast(T, import_string(value))
        except ImportError as e:
            raise errors.PyObjectError(error_message=str(e))


if TYPE_CHECKING:
    PyObject = Type # instead of `Callable`
else:
    class PyObject(Generic[T], _PyObject[T]):
        pass
```

This is fully backwards compatible with the current `PyObject` signature, the only adjustment would be using `PyObject[Any]` instead of the plain `PyObject`.

Attaching screenshot of working editor suggestions in VS Code:

![immagine](https://user-images.githubusercontent.com/30410534/154796951-18400541-c597-45b7-aff7-3457911d75b7.png)

This also suggests constructor arguments properly:

![immagine](https://user-images.githubusercontent.com/30410534/154796989-99832184-0bfe-4e20-8c1f-9b02a08206a7.png)

The other change I made here was to move the `PyObject` definition from the "else" branch to the actual codebase. The reason for this is that users could then be able to subclass `_PyObject`, should they need to customize the validation internals.
Currently, when trying to subclass `PyObject`, static checkers would complain because the actual implementation is not available at type checking time.

**Comment by dmontagu** [comment]
If I understand correctly, I think you can get the behavior you want via `cls: typing.Type[MyClass] = MyClass`. Does `PyObject` have some different behavior to this that is important to preserve? (`PyObject` has been removed in v2)

**Reply by rsayn** [reply]
Hey! Thanks for the heads-up. `PyObject` was useful when parsing configuration files into Pydantic classes, so it also used to import the declared type / object from a Python module when loading the file.

I'm not 100% sure this works the same with V2, I'll test it ASAP and let you know

For example:

```python
class MyConfig:
    some_string: str
    some_module: PyObject
    some_class: PyObject
    some_object: PyObject
```

```yaml
# config file
some_string: example
some_module: my_lib.package
some_class: my_lib.package.MyClass
some_object: my_lib.package.my_object
```

```python
# loading the config
>>> config = MyConfig(**config_dict_from_yaml)
>>> config.some_string
'example'
>>> config.some_module
<module 'my_lib.package' from [...]>
>>> config.some_class
<class 'my_lib.package.MyClass'>
```

**Reply by lovetheguitar** [reply]
Exactly our use case as well. It works very well for all kind of objects, e.g. also to load a list of instances from a module, what otherwise would be quite cumbersome, IIRC.

**Reply by rsayn** [reply]
Here's a couple snippets reproducing the issue:

`main.py`
```python
import json
from pathlib import Path
from types import ModuleType
from typing import Type
from pydantic import BaseModel



class MySettings(BaseModel, arbitrary_types_allowed=True):
    some_string: str
    some_module: ModuleType
    some_class: Type
    some_object: object


if __name__ == '__main__':
    settings_from_json = json.loads(Path("sample.json").read_text())
    settings = MySettings(**settings_from_json)

    print(settings.dict())
```

`sample.json`
```json
{
    "some_string": "example",
    "some_module": "typing",
    "some_class": "collections.deque",
    "some_object": "math.inf"
}
```

Output:

```
python main.py
Traceback (most recent call last):
  File "main.py", line 18, in <module>
    settings = MySettings(**settings_from_json)
  File "/Users/riccardosayn/dev/rsayn/pydantic-v2-tests/.venv/lib/python3.8/site-packages/pydantic/main.py", line 231, in __init__
    __pydantic_self__.__pydantic_validator__.validate_python(data, self_instance=__pydantic_self__)
pydantic_core._pydantic_core.ValidationError: 2 validation errors for MySettings
some_module
  Input should be an instance of module [type=is_instance_of, input_value='typing', input_type=str]
some_class
  Input should be a type [type=is_type, input_value='collections.deque', input_type=str]
```

**Reply by dmontagu** [reply]
It looks to me like what you want may have moved to `pydantic.types.ImportString`:
```python
from types import ModuleType
from typing import Type
from pydantic import BaseModel, ImportString


class MySettings(BaseModel, arbitrary_types_allowed=True):
    some_string: str
    some_module: ImportString[ModuleType]
    some_class: ImportString[Type]
    some_object: ImportString[object]


if __name__ == '__main__':
    settings = MySettings(**{
        "some_string": "example",
        "some_module": "typing",
        "some_class": "collections.deque",
        "some_object": "math.inf"
    })
    """
    pydantic_core._pydantic_core.ValidationError: 1 validation error for MySettings
    some_module
      Invalid python path: "typing" doesn't look like a module path [type=import_error, input_value='typing', input_type=str]
    """
    print(settings.dict())
```

**Reply by dmontagu** [reply]
However, I will note that it doesn't seem to allow strings without a `.` for module paths. I'm not sure why this is the case; if it's a problem for you, please create an issue.

**Comment by adriangb** [comment]
> This is fully backwards compatible with the current PyObject signature, the only adjustment would be using PyObject[Any] instead of the plain PyObject.

I'm not super familiar with the usage of PyObject but I'll point out that you can make it _fully_ backwards compatible by using:

```python
from typing_extensions import TypeVar

T = TypeVar('T', default=Any)
```

We plan on supporting this and type checkers will understand it.

**Reply by samuelcolvin** [reply]
That doesn't cover the use case because of processing strings as import paths.

**Comment by samuelcolvin** [comment]
We should allow imports of strings with no path.

But there's another problem:

When importing `foo.bar`, is `bar` a submodule, or an attribute of `foo`?

We should therefore switch to the syntax uvicorn uses: `foo:bar` - using `:` to identify attributes vs. modules.

If we make this change, we have to make it before V2 release.

**Reply by samuelcolvin** [reply]
I've created #5718 to track changing this.
